### DISCUSSION TITLE
Choice of DB to store our webhooks

### DISCUSSION THREAD

**Discussion by sidvenu** [thewebhookcompany/thewebhookcompany#9] [Ideas]
The [ingestor has been built](https://github.com/thewebhookcompany/thewebhookcompany/pull/2), and the accumulator is to be worked on soon.

For those who are new, we are terming the ingestor to be the service that ingests all incoming webhooks into a Kafka cluster, that is then read later by the accumulator to be inserted into a DB + further processing. Planning to write a README about the planned components of thewebhookcompany.

We need to think more about this DB with the context of scale we are thinking about.

### What scale are we thinking about?
This is a B2B open source product - our users are businesses themselves, potentially getting thousands or tens of thousands of customers each day. 1 relatively big business using this service could have the scale of 1000 small business using it, in terms of # of webhooks consumed per day. Our architecture should theoretically be able to handle 10x the peak load we are expecting.

When estimating the scale we need to plan out our architecture for, one major source of webhook I saw was Payments - [UPI processed 7.82B payments in Dec 2022](https://www.outlookindia.com/business/upi-transactions-rise-over-7-82-billion-in-december-2022-here-s-how-upi-has-grown-over-the-years-news-250301), translating to average ~3,000 transactions per second, which translates to maybe ~15-30,000 transactions per second on a peak level. [Visa famously claims](https://www.visa.co.in/run-your-business/small-business-tools/retail.html) to have the ability to process ~24,000 transactions per second at its peak. Let's say each transaction on average generates 2 webhooks. That's  ~78-108,000 webhooks per second (peak load) between UPI and Visa alone.

Our service probably won't reach this scale until massive adoption comes through at all levels of business - maybe a few years. Details like the number of servers we are using, specs of each server, etc. can be set as we go, but fundamental details like the kind of DB we are going to be using should take into account the potentially massive scale this service may achieve. Taking this into account, I believe the stack architecture we should be designing should theoretically handle maybe **_100-200,000 webhooks per second_** at its peak. Not the resource allocation, no, but the architecture should theoretically support that figure given enough (viable) resources.

### What are we going to use this DB for?
For now, probably a table for storing each webhook-attempt (including the webhook data, in the good old denormalised way). It will have rows with expiry times set between 3-7 days (we cannot store webhook data forever, it would be both pointless and very expensive). This expiry time will be decided on a per-user level.

### DB Shortlist
We need something with AP of CAP - Availability and Partition Tolerance. Strict Consistency is not a requirement. Something like HBase is out since it is CP. **_Cassandra_** seemed to be a good option since it has probably all we are looking for - eventual consistency that can be tweaked, support for storing a huge amount of data, horizontal scaling, but then I came across **_ScyllaDB_** through [Discord Engineering Team's blog post](https://discord.com/blog/how-discord-stores-trillions-of-messages).

It seems to be worthy to try out, but there is one problem. Expiring rows would create tombstones. Querying tombstones will be expensive. There is a solution - for us to design our system in a way that all our queries query only the latest N-1 days, with N being the expiry # of days of that user this webhook was being sent to. And of course, our partition key would include some sort of time parameter. This way, we'd avoid the partitions where tombstones are likely found. Of course, if we increase the expiry time for a particular user from 3 days to 7 days for example, we'd then query the last 6 days data of that user from then on. Which would lead to querying tombstones indeed. But such events are rare. Perhaps we can [force a delete of tombstones](https://docs.scylladb.com/stable/kb/tombstones-flush.html) in that case before updating the expiry time.

---

I'd love to hear your thoughts on the above. This writeup is not very formal, just my (possibly unorganised) thoughts and research out through text.

cc @shyam-king @ankit-rawani

**Comment by ankit-rawani** [comment]
ScyllaDB does seems like a really good option to go with. Will try to check it out in detail.
Another option I feel we can look into is influxDB. It's a time series database so might be a good fit for our application. Though I feel Scylla has more advantages over influx like higher throughput, lower latency etc.

**Reply by sidvenu** [reply]
Wait why do you think a time series DB would fit this usecase?

**Reply by ankit-rawani** [reply]
Like I feel there is a need to store the webhooks based on time and query it based on the dropoff time. Also, the data would be supposed to be kept only for certain amount of time which like a default behaviour in influxdb. 

But ye after going through Scylla for past few days, I feel Scylla is much better option.

**Comment by sidvenu** [comment]
Coming back from #11 

Our only requirement right now is to allow for folks to search through the webhook datastore to see if they received a webhook, the retries data etc. I see OpenSearch and Apache Solr as a strong contenders for this. What do you think?

**Comment by sidvenu** [comment]
From what I see, OpenSearch looks great. Lets go with that for now and see how it goes.
